sys/include/string_utils.h

```diff
@@ -38,6 +39,67 @@
 extern "C" {
 #endif
 
+/**
+ * @brief   String Writer structure.
+ *          Helper for writing multiple formatted strings to a buffer
+ */
+typedef struct {
+    const char *start;  /**< start of the target buffer */
+    char *position;     /**< current write pointer      */
+    size_t capacity;    /**< remaining capacity of the buffer */
+} string_writer_t;
+
+/**
+ * @brief   Initialize a string writer structure
+ *
+ * @param[out]  sw      String Writer object to initialize
+ * @param[in]   buffer  destination buffer
+ * @param[in]   len     size of the destination buffer
+ */
+static inline void string_writer_init(string_writer_t *sw, void *buffer, size_t len)
+{
+    assert(buffer && len);
+
+    sw->start = buffer;
+    sw->position = buffer;
+    sw->capacity = len;
+    sw->position[0] = 0;
+}
+
+/**
+ * @brief   Get the size of the string contained by the string writer
+ * @param[in]   sw      String Writer to query
+ * @return      size of the string
+ */
+static inline size_t string_writer_len(const string_writer_t *sw)
+{
+    return sw->position - sw->start;
+}
+
+/**
+ * @brief   Get the string contained by the string writer
+ * @param[in]   sw      String Writer to query
+ * @return      the string assembled by string writer
+ */
+static inline const char *string_writer_str(const string_writer_t *sw)
+{
+    return sw->start;
+}
+
+/**
+ * @brief   Write a formatted string to a buffer
+ *          The string will be truncated if there is not enough space left in
+ *          the destination buffer.
+ *          A terminating `\0` character is always included.
+ *
+ * @param[in]   sw      String Writer to write to
+ * @param[in]   format  format string to write
+ *
+ * @return      number of bytes written on success
+ *              -E2BIG if the string was truncated
+ */
+int swprintf(string_writer_t *sw, const char *restrict format, ...);
```

Comment:
you can add
```c
__attribute__ ((format (printf, 2, 3)))
```
to static printf-like compiler checks for the format string and arguments, see https://gcc.gnu.org/onlinedocs/gcc-3.2/gcc/Function-Attributes.html.

Reply:
Odd, I tried now and works for me:
```c
__attribute__ ((format (printf, 2, 3)))
int __swprintf(string_writer_t *sw, FLASH_ATTR const char *restrict format, ...);
```
Then:
```c
    res = swprintf(&sw, "Hello World!");
```
works just fine, while:
```c
    res = swprintf(&sw, "Hello World%s!");
```
results in 
```
/home/mihai.renea@ml-pa.loc/proj/RIOT/tests/unittests/tests-libc/tests-libc.c: In function ‘test_libc_s
wprintf’:
/home/mihai.renea@ml-pa.loc/proj/RIOT/tests/unittests/tests-libc/tests-libc.c:42:38: error: format ‘%s’
 expects a matching ‘char *’ argument [-Werror=format=]
   42 |     res = swprintf(&sw, "Hello World%s!");
      |                                     ~^
      |                                      |
      |                                      char *
```

Which compiler version are you using? Mine:
```
$ arm-none-eabi-gcc --version
arm-none-eabi-gcc (15:10.3-2021.07-4) 10.3.1 20210621 (release)
```

Reply:
This is only happens with `avr-gcc (GCC) 7.3.0` 

```
/home/benpicco/dev/RIOT/tests/unittests/tests-libc/tests-libc.c: In function ‘test_libc_swprintf’:
/home/benpicco/dev/RIOT/tests/unittests/tests-libc/tests-libc.c:42:5: error: format not a string literal and no format arguments [-Werror=format-security]
     res = swprintf(&sw, "Hello World!");
     ^~~
/home/benpicco/dev/RIOT/tests/unittests/tests-libc/tests-libc.c:49:5: error: format not a string literal and no format arguments [-Werror=format-security]
     res = swprintf(&sw, "0123456789");
     ^~~
/home/benpicco/dev/RIOT/tests/unittests/tests-libc/tests-libc.c:54:5: error: format not a string literal and no format arguments [-Werror=format-security]
     res = swprintf(&sw, "01234567891");
     ^~~
/home/benpicco/dev/RIOT/tests/unittests/tests-libc/tests-libc.c:59:5: error: format not a string literal and no format arguments [-Werror=format-security]
     res = swprintf(&sw, "###");
     ^~~
```

Reply:
Ah I think it's because the `TO_FLASH()` is interfering